### ISSUE TITLE
"The checker shouldn't produce errors on unannotated files"

### ISSUE THREAD

**Issue by cpovirk** [jspecify/jspecify#77]
(That is: It might produce _warnings_, which could fail analysis in "strict mode." But "lenient mode" should succeed.)

This is a principle we've discussed a couple times ([link](https://github.com/google/codeanalysis-annotations/issues/45#issuecomment-546054561), [link](https://github.com/google/codeanalysis-annotations/issues/45#issuecomment-546266834)). From the second link:

> I think this should follow directly from the fact that `@NullUnspec` as we define it communicates zero semantic nullness information. So if a checker fails for some reason, it cannot be due to information it got from our annotations, because there was no such information.

However: While a file might not be annotated, it might use APIs from other files that are. For example, assuming that `Optional` is annotated:

```java
[@DefaultNullnessUnspecified]
class Foo {
  void bar() {
    Optional.of(null); // error, presumably?
  }
}
```

That's "only" implementation code. Still, it's worth keeping in mind that we may have a cleanup to do before we can enable implementation checking universally.

Alternatively, we could consider making the checker skip `@DefaultNullnessUnspecified` APIs -- or just demote errors to warnings? But probably we'd want to roll out the checking gradually instead.

That all seems fine, hopefully? Just something that I'd overlooked initially.

**Comment by cpovirk** [comment]
It occurs to me that checkers can find errors even in code that's entirely reasonable:

```java
map.get(foo).dereferenceIt();
```

With plain jspecify annotations, that is very likely to be an error.

I still think that's fine. It's just a reminder to us at Google (and perhaps other organizations) that turning on naive null checking _everywhere_ would require a ton of suppressions to avoid breaking builds.

**Comment by kevin1e100** [comment]
.... hence why there are those that are lobbying to let `Map.get` return
unspecified nullness :). [ this is one reason for doing that anyway ]

> With plain jspecify annotations, that is very likely to be an error.

I believe the intention was to leave it up to checkers to decide what to
do?  I didn't think we would specify this to be an error for sure, or am I
missing something?

Nonetheless the point is certainly valid that there *will* be places where
annotations would suggest errors in unannotated code, and this will present
challenges for tools.  One response, of course, being not to do anything by
default until the file itself is annotated.

**Comment by cpovirk** [comment]
Sorry, yes, I continue to say "error" as shorthand for "type mismatch if the only information available to you is that from jspecify annotations," but I should stop doing that.

And sorry, this whole issue is more about adoption of tools than it is about our annotations per se. I just wanted to make sure we were aware that rolling out universal nullness checking _will_ produce such "mismatches" in unannotated code, and in fact it's likely to be common. Even if we use a tool that has smarts for `Map.get` specifically, this same problem is bound to arise frequently (as I believe you've said we've seen with Android SDK upgrades :))

Maybe I should just close this issue, since I think we all agree on the facts. (Maybe we always did; I just thought I once detected an undercurrent to the contrary.) Given that, the issue is serving mainly as a forum for me to put my foot in my mouth :)

**Comment by cpovirk** [comment]
@kevin1e100 , @netdpb , and I had discussed this a bit last week, coming to similar conclusions as above:
- Only implementation code can break. (Reasoning: Annotations on APIs can't be wrong if the APIs are unannotated :) In contrast, _implementations_ can be broken because they can dereference `@Nullable` return types, etc.)
- Fixes for broken implementation code can be made locally (without having to also update your own callers). (At the least, you have the option to suppress the warning, leaving your signature untouched. Other times, you might add a null check or something. Sometimes you will want to change your signature, but you don't _need_ to immediately.)

But I did just realize that there is an edge case in which APIs could break:

```java
@NullAware
interface Super1 {
  List<String> foo();
}

@NullAware
interface Super2 {
  List<@Nullable String> foo();
}

interface Sub extends Super1, Super2 {}
```

Tools could defensibly produce an error for `Sub`, since it [inherits incompatible signatures](https://github.com/jspecify/jspecify/issues/68) for `foo()`. This is true even though:

- It is unannotated code.
- It doesn't even declare the problematic method directly.

I would imagine that this case is very rare. And it should come up only when someone adds `@Nullable` to a type in the supertype -- a change that is already potentially a breaking change and thus one that should be tested accordingly. So this may all be academic.
